### PR TITLE
feat: migrate aka github/twitter links

### PR DESCRIPTION
**Pull request by zachferland**

### CODE REVIEW

**Comment by zachferland** [comment]
Was getting timeouts on writing to aka definition, have not looked in to whats causing that or missing. Otherwise this is the implementation. Also will add a simple test for this once resolved.

File: package.json [hunk 1]
```diff
@@ -53,15 +53,16 @@
...
+    "@ceramicstudio/idx-constants": "^0.7.0",
     "@ethersproject/hdnode": "^5.0.8",
     "@stablelib/sha256": "^1.0.0",
     "3id-did-provider": "^0.5.1",
     "caip": "^0.9.2",
     "dag-jose": "^0.3.0",
-    "dids": "^1.1.0",
+    "dids": "^1.1.1",
     "errors-utils": "^0.1.0",
+    "jwt-decode": "^3.1.2",
```
Comment (oed) [inline package.json:65]: We don't need this. Just use `uint8arrays`.

File: src/connectService.ts [hunk 2]
```diff
@@ -249,25 +260,97 @@ class ConnectService extends IframeService {
...
+
+  async migrateAKALinks(did: string, profile = {} as any): Promise<void> {
     assert.isDefined(this.idx, 'IDX instance must be defined')
     assert.isDefined(this.ceramic, 'Ceramic instance must be defined')
-    assert.isDefined(this.threeId, 'ThreeIdProvider instance must be defined')
+    assert.isDefined(this.provider, 'ThreeIdProvider instance must be defined')
 
-    const linkProof = await get3BoxLinkProof(did)
-    if (linkProof) await this._writeLinkProof(accountId, linkProof)
+    const dids = new DID({
```
Comment (oed) [inline src/connectService.ts:330]: I would name this variable `user` for clarity :)

File: src/connectService.ts [hunk 3]
```diff
@@ -249,25 +260,97 @@ class ConnectService extends IframeService {
...
+      const tweetUrl = decoded.claim?.twitter_proof
+      const challengeCode = await linkRequest(type, did, twitterHandle)
+      const jws = await dids.createJWS({ challengeCode })
+      const twitterProof = await linkVerify(type, jws, tweetUrl)
+      return {
+        protocol: 'https',
+        host: 'twitter.com',
+        id: twitterHandle,
+        claim: tweetUrl,
+        attestations: [{ 'did-jwt': twitterProof }],
```
Comment (oed) [inline src/connectService.ts:296]:
```suggestion
        attestations: [{ 'did-jwt-vc': twitterProof }],
```

File: src/connectService.ts [hunk 4]
```diff
@@ -249,25 +260,97 @@ class ConnectService extends IframeService {
...
+      if (!githubHandle) throw new Error('link fail')
+      const challengeCode = await linkRequest(type, did, githubHandle)
+      const jws = await dids.createJWS({ challengeCode })
+      const githubProof = await linkVerify(type, jws, gistUrl)
+      return {
+        protocol: 'https',
+        host: 'github.com',
+        id: githubHandle,
+        claim: gistUrl,
+        attestations: [{ 'did-jwt': githubProof }],
```
Comment (oed) [inline src/connectService.ts:318]:
```suggestion
        attestations: [{ 'did-jwt-vc': githubProof }],
```

**Comment by zachferland** [comment]
All set here now.

**Review by oed — APPROVED** [review]

File: src/connectService.ts [hunk 5]
```diff
@@ -46,6 +55,13 @@ const rpcError = (id: string | number) => {
   return Object.assign(rpcError.toObject(), { id })
 }
 
+const jwt_decode = <T>(jwt: string): T => {
```
Comment (PaulLeCam) [inline src/connectService.ts:58]: Could be named `decodeJWT` maybe to be more consistent with other function names?

File: src/connectService.ts [hunk 6]
```diff
@@ -169,8 +184,9 @@ class ConnectService extends IframeService {
     this.setActiveAccount(accountId)
 
     if (migration && legacyDid) {
-      await this.migrate3BoxProfile(legacyDid)
+      const profile3Box = await this.migrate3BoxProfile(legacyDid)
       await this.migrate3BoxLinks(legacyDid, accountId)
+      await this.migrateAKALinks(legacyDid, profile3Box)
```
Comment (PaulLeCam) [inline src/connectService.ts:189]: Could these 2 calls run in parallel?

File: src/connectService.ts [hunk 7]
```diff
@@ -249,25 +266,97 @@ class ConnectService extends IframeService {
...
+  async migrate3BoxProfile(did: string): Promise<any> {
     assert.isDefined(this.idx, 'IDX instance must be defined')
-    assert.isDefined(this.ceramic, 'Ceramic instance must be defined')
-    assert.isDefined(this.threeId, 'ThreeIdProvider instance must be defined')
-
     const profile = await get3BoxProfile(did)
     const transform = transformProfile(profile)
     const existing = (await this.idx.get('basicProfile')) || {}
-    console.log(Object.assign(existing, transform))
     await this.idx.set('basicProfile', Object.assign(existing, transform))
```
Comment (PaulLeCam) [inline src/connectService.ts:275]: `await idx.merge('basicProfile', transform)` could be used instead here, as it implements this logic of load and shallow merge

File: src/connectService.ts [hunk 8]
```diff
@@ -249,25 +266,97 @@ class ConnectService extends IframeService {
...
+
+    const results: Array<
+      Array<AlsoKnownAsAccount> | AlsoKnownAsAccount | null
+    > = await Promise.all([
+      existing(this.idx),
+      this._twitterVerify(user, did, profile),
+      this._githubVerify(user, did, profile),
+    ])
+
+    type ExcludesBoolean = <T>(x: T | null) => x is T
```
Comment (PaulLeCam) [inline src/connectService.ts:354]: This type definition could be set at the module level, no?

**Review by PaulLeCam — APPROVED** [review]
Logic looks good to me, I just noticed a few small things but don't hesitate to ignore, should be good to merge as-is.

File: src/connectService.ts (hunk 6 above)
Comment (zachferland) [reply src/connectService.ts:189]: yeah, good call

File: src/connectService.ts (hunk 7 above)
Comment (zachferland) [reply src/connectService.ts:275]: oh nice, yeah will start using that instead

**Pull request merged**